Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm

```diff
@@ -0,0 +1,178 @@
+#import <wtf/RetainPtr.h>
+
+static void testRestoreLocalStorage(RetainPtr<WKWebView> webView)
+{
+    auto navigationDelegate = adoptNS([[TestNavigationDelegate alloc] init]);
+    [webView setNavigationDelegate:navigationDelegate.get()];
+    [webView loadHTMLString:@"<body>webkit.org</body>" baseURL:[NSURL URLWithString:@"https://webkit.org"]];
+    [navigationDelegate waitForDidFinishNavigation];
+
+    __block bool doneEvaluatingJavaScript = false;
+    [webView evaluateJavaScript:@"localStorage.setItem('key', 'value');" completionHandler:^(id value, NSError *error) {
+        EXPECT_NULL(error);
+        doneEvaluatingJavaScript = true;
+    }];
+    TestWebKitAPI::Util::run(&doneEvaluatingJavaScript);
+
+    RetainPtr set = [NSSet setWithObject:WKWebsiteDataTypeLocalStorage];
+
+    __block bool doneFetching = false;
+    [[[webView configuration] websiteDataStore] _fetchDataOfTypes:set.get() completionHandler:^(NSData *data) {
+        EXPECT_NOT_NULL(data);
+        EXPECT_GT([data length], 0u);
+        doneFetching = true;
+    }];
+    TestWebKitAPI::Util::run(&doneFetching);
+}
+
+TEST(WebKit, RestoreLocalStoragePersistentDataStore)
+{
+    testRestoreLocalStorage(adoptNS([WKWebView new]));
+}
+
+TEST(WebKit, RestoreLocalStorageEphemeralDataStore)
+{
+    RetainPtr storeConfiguration = adoptNS([[_WKWebsiteDataStoreConfiguration alloc] initNonPersistentConfiguration]);
+    RetainPtr viewConfiguration = adoptNS([WKWebViewConfiguration new]);
+    [viewConfiguration setWebsiteDataStore:adoptNS([[WKWebsiteDataStore alloc] _initWithConfiguration:storeConfiguration.get()]).get()];
+    RetainPtr webView = adoptNS([[WKWebView alloc] initWithFrame:CGRectMake(0, 0, 100, 100) configuration:viewConfiguration.get()]);
+
+    testRestoreLocalStorage(webView);
+}
+
+@interface RestoreLocalStorageMessageHandler : NSObject <WKScriptMessageHandler>
+@end
+
+@implementation RestoreLocalStorageMessageHandler
+
+- (void)userContentController:(WKUserContentController *)userContentController didReceiveScriptMessage:(WKScriptMessage *)message
+{
+    receivedScriptMessage = true;
+    lastScriptMessage = message;
+}
+
+@end
+
+static NSString *mainFrameString = @"<script> \
+    function postResult(event) \
+    { \
+        window.webkit.messageHandlers.testHandler.postMessage(event.data); \
+    } \
+    addEventListener('message', postResult, false); \
+    </script> \
+    <iframe src='https://127.0.0.1:9091/'>";
+
+static constexpr auto frameBytes = R"TESTRESOURCE(
+<script>
+function postMessage(message)
+{
+    parent.postMessage(message, '*');
+}
+async function setItem()
+{
+    try {
+        localStorage.setItem('key', 'value');
+        postMessage('Item is set');
+    } catch(err) {
+        postMessage('error: ' + err.name + ' - ' + err.message);
+    }
+}
+setItem();
+</script>
+)TESTRESOURCE"_s;
+
+enum DataStoreType {
+    Persistent,
+    Ephemeral
+};
+
+static void testRestoreLocalStorageThirdPartyIFrame(DataStoreType dataStoreType)
```

Comment:
This function can directly take `RetainPtr<WKWebsiteDataStore>`, and you could construct store before it; no need for the enum